src/adapter/src/optimize/dataflow_expiration.rs

```diff
@@ -0,0 +1,212 @@
+    ) -> TimeDependence {
+        use TimeDependence::*;
+
+        // Collect all time dependencies of our dependencies.
+        let mut time_dependencies = ids
+            .into_iter()
+            .map(|id| self.determine_dependence_inner(id))
+            .collect::<Vec<_>>();
+
+        // Sort and dedupe to remove redundancy.
+        time_dependencies.sort();
+        time_dependencies.dedup();
+
+        let mut time_dependence = if time_dependencies.iter().any(|dep| matches!(dep, Wallclock)) {
+            // Wall-clock dependency is dominant.
+            RefreshSchedule(schedule, vec![Wallclock])
+        } else if time_dependencies
+            .iter()
+            .any(|dep| matches!(dep, RefreshSchedule(_, _)))
+        {
+            // No immediate wall-clock dependency, found some dependency with a refresh schedule.
+            // Remove remaining Indeterminate dependencies.
+            time_dependencies.retain(|dep| matches!(dep, RefreshSchedule(_, _)));
+            RefreshSchedule(schedule, time_dependencies)
+        } else {
+            // No wall-clock dependence, no refresh schedule
+            Indeterminate
+        };
+        time_dependence.normalize();
+        time_dependence
+    }
+
+    /// Determine the time dependence for a [`DataSourceDesc`].
+    ///
+    /// Non-epoch timelines are indeterminate, and load generators.
+    fn for_data_source_desc(
+        &mut self,
+        desc: &DataSourceDesc,
+        timeline: &Timeline,
+    ) -> TimeDependence {
+        use TimeDependence::*;
+        // We only know how to handle the epoch timeline.
+        if !matches!(timeline, Timeline::EpochMilliseconds) {
+            return Indeterminate;
+        }
+        match desc {
+            DataSourceDesc::Ingestion { ingestion_desc, .. } => {
+                match ingestion_desc.desc.connection {
+                    // Kafka, Postgres, MySql sources follow wall clock.
+                    GenericSourceConnection::Kafka(_)
+                    | GenericSourceConnection::Postgres(_)
+                    | GenericSourceConnection::MySql(_) => Wallclock,
+                    // Load generators not further specified.
+                    GenericSourceConnection::LoadGenerator(_) => Indeterminate,
+                }
+            }
+            DataSourceDesc::IngestionExport { ingestion_id, .. } => {
+                self.determine_dependence_inner(*ingestion_id)
+            }
+            // Introspection, progress and webhook sources follow wall clock.
+            DataSourceDesc::Introspection(_)
+            | DataSourceDesc::Progress
+            | DataSourceDesc::Webhook { .. } => Wallclock,
+        }
+    }
+
+    /// Determine the time dependence for a single global ID.
+    fn determine_dependence_inner(&mut self, id: GlobalId) -> TimeDependence {
+        use TimeDependence::*;
+
+        if let Some(dependence) = self.seen.get(&id).cloned() {
+            return dependence;
```

Comment:
Just an observation, on the `apply()` side, these common subtrees will be traversed multiple times. If this ever becomes a problem, we could preserve the ids and apply this `seen` logic even there.